Adding support for @implements.

dragomirtitian opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #35629


Open issues:
1. [x] How to handle interfaces that can't be implemented in JS, for example those with index signatures
2. [x] Fix code fix that implements missing interface members (currently it does not show up, but it would be nice if it worked for JS as well)
3. [x] Add more tests

@sandersn For 1 I would suppress these errors for JS.  For 2, do you think it is worth showing the fix?  Also maybe we could have a typescript-bot pack of this so @TimvdLippe can maybe give it a try. 

One issue, unrelated to `@implements` necessarily, is that method in the classes used as interfaces in @TimvdLippe example don't have any return statement, so any return type will cause an error. Support for `@abstract` was taken out of the scope of #35629, but maybe adding support for `@abstract` could be useful to suppress these errors. Using `@ts-ignore` might also be a simpler workaround.

TimvdLippe commented:
Yes I would like to try it out. I will be able to do so next week when I am back in the office.

TimvdLippe commented:
To be able to debug, I would need a bot-pack build that I can try out. Let me know when you have one available :smile:

dragomirtitian commented:
@TimvdLippe I unfortunately can't summon the TS bot .. let me try something, <clears throat />  

Oh mighty and benevolent TS overlords @sandersn  @orta  or @DanielRosenwasser  could we have a pack of this ?

orta commented:
Sure!
@typescript-bot pack this

sandersn commented:
@dragomirtitian re your questions:
1. Suppressing errors that can't be fixed in JS is a good idea. We can re-instate them later if we come up with a way to fix them.
2. I believe the code you emit for the code fix would have to be different, at least in that it uses jsdoc instead of type annotations. And it may have to fail because of errors from (1). That is a lot of work so you can open a new issue instead and assign it to me (or ping me on it).

I will look at the code now.

I'm a linguist, so I prefer the traditional 3 invocations. Like this: @noamchomsky @noamchomsky @noamchomsky

sandersn commented on src/compiler/types.ts:
```diff
@@ -2638,11 +2639,11 @@ namespace ts {
      */
-    export interface JSDocAugmentsTag extends JSDocTag {
-        kind: SyntaxKind.JSDocAugmentsTag;
+    export interface JSDocHeritageTag extends JSDocTag {
```
this is a little surprising. I expected to see a new type JSDocImplementsTag and then maybe
`type JSDocHeritageTag = JSDocAugmentsTag | JSDocImplementsTag` for convenience.

dragomirtitian replied:
I could create two separate types, but they happen to have the exact same structure, since `@implements` in Jsdoc takes one type, but can appear multiple times (at least as far as I have read and seen). I thought it made sense to unify all of them since they kind of are related.

sandersn replied:
Hmmm, I still think it would be less confusing to have two separate types since that's how the other types work.

Also: good point, I didn't think about multiple `@implements`. Did you have tests for that already?

dragomirtitian replied:
Ok, I will split up the types. I did do some testes and multiple `@implement` tags work, but I did not add the tests to git yet, I will do so this weekend.

sandersn commented on src/compiler/utilities.ts:
```diff
@@ -2769,6 +2769,14 @@ namespace ts {
         return heritageClause && heritageClause.types.length > 0 ? heritageClause.types[0] : undefined;
     }
 
+    export function getEffectiveImplementsTypeNodes(node: ClassLikeDeclaration): undefined | readonly ExpressionWithTypeArguments[]{
```
it would be a lot safer to put this code into getClassImplementsHeritageClauseElements, and then change the name to `getEffectiveImplementsTypeNodes`.

sandersn commented on src/compiler/utilitiesPublic.ts:
```diff
@@ -787,7 +792,12 @@ namespace ts {
+        return getJSDocTags(node).filter(predicate);
+    }
+
+    /** Gets all JSDoc tags of a specified kind */
```
I wish we could rid of getAllJSDocTagsOfKind now &mdash; we don't actually use it, but it's part of the public API.

dragomirtitian replied:
I removed the `or undefined` part since it returns an empty array never `undefined`. Should I mark `getAllJSDocTags` as internal so as not to expand the public API ?

sandersn replied:
No, getAllJSDocTags is better, I think people will be able to figure out which to use based on the type.

TimvdLippe commented:
I can confirm that the updated build is catching issues like missing overrides. That said, we are hitting a couple of other issues with TypeScript (https://github.com/microsoft/TypeScript/issues/19983) which makes it hard for us to fully test all of its internals. My understanding thus far is that this change is WAI.

One improvement that would be great is that methods with `@return` on an `@interface` do not result in the `A function whose declared type is neither 'void' nor 'any' must return a value.` error. At that point, I don't see any TypeScript blockers for supporting JSDoc interfaces (without special-casing for TypeScript).

Thank you for doing the work and for the quick implementation!

dragomirtitian commented:
@TimvdLippe  The scope of this was specifically defined as support for `@implements` but I saw the same issue as you with regard to the actual usability of this with functions using `@return`, and I mentioned it in my opening comment to this pr. If @sandersn gives me the go ahead to suppress the error for classes marked with `@interface` I can roll up that chage in this pr, no problem (would probably have it done by Monday). 

If not I think the only workaround for now would be to `@ts-ignore` it, although I know that is not really a viable solution for this existing code base...

sandersn commented:
@dragomirtitian I'd like to keep support for `@interface` in a separate PR. Anyway, I'm not keen on the idea of ignoring errors on what would be invalid Typescript.
@TimvdLippe Is `// @ts-nocheck` at the beginning of the interface file a solution? I assume that it will be replaced by a d.ts once you are further along in the migration.

TimvdLippe commented:
@sandersn Given our extensive use of `@interface`, a no-check would probably not fly. Especially as the interfaces and usages are scattered throughout the codebase. I suppose adding `ts-ignore` for every single method is the least intrusive (interim) solution.

I am okay with punting that to a different PR though.

sandersn commented:
Looking at the example in [the jsdoc documentation](https://jsdoc.app/tags-interface.html) (excerpted below), I note that throwing in the interface body satisfies Typescript because throwing causes the function to return `never`, which is assignable to anything:

```js
/**
 * @interface
 */
function Color() {}

/**
 * @returns {Array<number>} An array containing the red, green, and blue values, in that order.
 */
Color.prototype.rgb = function() {
    throw new Error('not implemented');
};
```

Does that work with Closure?

TimvdLippe commented:
@sandersn I just checked and the following does indeed compile in both Closure and TS (with this patch):

```js
/**
 * @interface
 */
export class EventTarget {
  /**
   * @param {symbol} eventType
   * @param {function(!Common.Event)} listener
   * @param {!Object=} thisObject
   * @return {!Common.EventTarget.EventDescriptor}
   */
  addEventListener(eventType, listener, thisObject) {
    throw new Error(`Not implemented`);
  }
}
```

dragomirtitian commented:
@sandersn I fixed the code review issues. I added a test for multiple interfaces. I ran into the following issues;

1. I am not sure exactly how to suppress the error for signature implementation. I mean I know how the error is produced (the error is deep in the bowels of `checkTypeRelatedTo` which is called when checking the heritage clauses). My issue is I am unsure how to suppress errors coming from there. I could add a parameter to `checkTypeRelatedTo` to skip index checks during implements checks, but that seems very ad hoc. The other option might be to capture diagnostics using `errorOutputContainer` but I would have to see if that could work. If you can think of an example similar to this, I could use the 'inspiration' to make sure I don't go off the rails 😂.

2. You mentioned four-slash tests for type completions in `@implements` which would be similar to `@extends`. The issue is but I could not trigger it completions for `@augments` and `@extends` at all I would create a separate issue to fix both. Thoughts?

Also sorry for the delay, it's been a hectic week at work.

sandersn commented:
1. I couldn't think of a way to suppress errors that was both easy and obviously correct. Plus, after more thought, I'm not sure it's the right thing. `@implements` is essentially just a check, which means that people will only use it get errors, not to provide better intellisense. (Disregarding the completions that don't work and should.) For people who care mainly about correctness, having JS behave differently from TS is only going to confuse them. And people with random `@implements`
2. Definitely create a separate issue for completions.

sandersn commented:
@typescript-bot user test this

sandersn commented:
ha ha oh no now user tests are broken because of the new octokit.

sandersn commented:
(fix is at #36915; I'll merge it after I see that it can post a user test PR)

sandersn commented:
@dragomirtitian can you merge from master and then I'll run the user tests?

dragomirtitian commented:
@sandersn Yes, I'll merge from master. First I want to fix the declaration emit for `@interface` since it is not being emitted right now. (Forgot about declaration emit in JS)

dragomirtitian commented:
@sandersn I merged master, and fixed declaration emit. You should have a look at the emit fixes.

sandersn commented:
@typescript-bot user test this

dragomirtitian commented:
@sandersn The user tests seem to have [failed](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=65408&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=4d2ea636-6e6e-5831-8b5a-8ce5e9ce8bbc) again :(

sandersn commented on src/compiler/checker.ts:
```diff
@@ -8337,6 +8352,29 @@ namespace ts {
+            type.resolvedImplementsTypes = type.resolvedImplementsTypes || emptyArray;
+            for (const declaration of type.symbol.declarations) {
+
+                if (!isClassLike(declaration)) continue;
```
Is there a test of either fail case? I didn’t see one.

dragomirtitian replied:
The `isClassLike` test guards if there are multiple merged declarations, but in all fairness I can't think of way to trigger this. Since we are in a JS file this will always be a class. Only interfaces and namespaces would merge with the class (and be present in the `declarations` array) but since we don't support declaring these in JS files, there is no way to trigger it. (Merging with an interface from `ts` file would not be included in this `declarations` as this is just the local file symbol)

**Should I just remove the test ?** It is a more of a future proof, if `@interface` or `@namespace` are ever included, but it might be just best to let whoever implements that deal with this case.

As for the `!implementsTypeNodes` tests, while I did not add any new tests for it, there are existing tests for emitting class definitions from a js file, which all do not have any `@implements` so those tests cover this as well (ex: any class in `jsDeclarationsClasses.ts` would hit this condition)

sandersn replied:
I think you should remove this check too. In future, if constructor functions support `@implements`, I'd prefer to just update `getEffectiveImplementsTypeNodes` and not here too.

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -8063,6 +8067,17 @@ namespace ts {
         }
 
+        function getImplementsTypes(type: InterfaceType): BaseType[] {
+            if (!type.resolvedImplementsTypes) {
```
@sandersn I indeed based my `getImplementsTypes` on `getBaseTypes`, seemed right to follow the same pattern, but you are right the caching is not necessary and I removed it.

sandersn replied:
Did you push the changes already? I didn't see any new commits.

dragomirtitian replied:
Sorry long weekend with sick children. I'm almost done but didn't get a change to finish. I'll get to it later today.

sandersn replied:
Oh no. =( I can relate; please don't spend time on this until you're able.

dragomirtitian commented on src/compiler/checker.ts:
```diff
@@ -8063,6 +8067,17 @@ namespace ts {
 
+        function getImplementsTypes(type: InterfaceType): BaseType[] {
+            if (!type.resolvedImplementsTypes) {
+                if (type.symbol.flags & SymbolFlags.Class) {
```
Since `type` can be other things than a class based of  `flags`, I thought it best to add the assertion in case someone else uses it in some other context. At the moment I use it only from `serializeAsClass` where the `type` is already guaranteed to be a class by previous checks. 

Should I just remove the check?

sandersn replied:
At the moment, you're basically just checking that `getDeclaredTypeOfClassOrInterface` returns a class -- and nothing in the code itself actually relies on the type being a class. I'd remove it. (And In case `getImplementsTypes` gets used elsewhere, I think usage is pretty obvious.)

sandersn commented:
I think this will be ready once you get rid of the unneeded checks.

dragomirtitian commented:
Ok, I think we are good to go!

sandersn commented:
@dragomirtitian Thanks for your hard work on this one!
